### PR TITLE
Do not consider MutableDocument's state for DocumentSnapshot comparison

### PR DESCRIPTION
**Pull request by ehsannas**
Hey there! So you want to contribute to a Firebase SDK?
Before you file this pull request, please read these guidelines:

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here.
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help
    us make Firebase APIs better, please propose your change in a feature request so that we
    can discuss it together.

### CODE REVIEW

File: Firestore/core/src/model/mutable_document.cc [hunk 1]
```diff
@@ -113,6 +113,11 @@ std::string MutableDocument::ToString() const {
   return stream.str();
 }
 
+bool MutableDocument::CompareIgnoringState(const MutableDocument& other) const {
```
Comment (schmidt-sebastian) [inline Firestore/core/src/model/mutable_document.cc:116]: If we keep this, this should be called "EqualsSomethingSomething" as it does not compare documents.

File: Firestore/core/src/api/document_snapshot.cc [hunk 2]
```diff
@@ -88,7 +88,12 @@ absl::optional<google_firestore_v1_Value> DocumentSnapshot::GetValue(
 bool operator==(const DocumentSnapshot& lhs, const DocumentSnapshot& rhs) {
   return lhs.firestore_ == rhs.firestore_ &&
          lhs.internal_key_ == rhs.internal_key_ &&
-         lhs.internal_document_ == rhs.internal_document_ &&
+         lhs.internal_document_.has_value() ==
+             rhs.internal_document_.has_value() &&
+         (lhs.internal_document_.has_value()
+              ? lhs.internal_document_->get().CompareIgnoringState(
```
Comment (schmidt-sebastian) [inline Firestore/core/src/api/document_snapshot.cc:94]: Nit (or personal preference): What do you think of comparing the value and the key from the document, and then using `exists()` and `snapshotMetadata` for the comparison of the remaining state (all encapsulated in this equals function).

File: Firestore/core/src/model/mutable_document.cc [hunk 3]
```diff
@@ -113,6 +113,11 @@ std::string MutableDocument::ToString() const {
...
+  return key_ == other.key_ && document_type_ == other.document_type_ &&
+         version_ == other.version_ && *value_ == *other.value_;
```
Comment (schmidt-sebastian) [inline Firestore/core/src/model/mutable_document.cc:118]: I think we might want to ignore version as well since it is also not exposed.

**Review by schmidt-sebastian — COMMENTED** [review]
Sorry for the slow review! Please ping me if you don't hear back within 24 hours (minus weekends - DO NOT PING ME ON WEEKENDS (*))

(*) at your own discretion.

**Comment by schmidt-sebastian** [comment]
You should also add a changelog entry.

File: Firestore/core/src/api/document_snapshot.cc (hunk 2 above)
Comment (ehsannas) [reply Firestore/core/src/api/document_snapshot.cc]: Good idea. Done.

**Review by schmidt-sebastian — APPROVED** [review]

File: Firestore/core/src/api/document_snapshot.cc [hunk 4]
```diff
@@ -88,7 +88,12 @@ absl::optional<google_firestore_v1_Value> DocumentSnapshot::GetValue(
 bool operator==(const DocumentSnapshot& lhs, const DocumentSnapshot& rhs) {
   return lhs.firestore_ == rhs.firestore_ &&
          lhs.internal_key_ == rhs.internal_key_ &&
-         lhs.internal_document_ == rhs.internal_document_ &&
+         lhs.exists() == rhs.exists() &&
+         (lhs.exists() ? lhs.internal_document_->get().key() ==
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply Firestore/core/src/api/document_snapshot.cc]: oops. I was checking the key twice (`lhs.internal_key_ == rhs.internal_key_` is already included).

**Pull request merged**